File: src/sandbox/backends/wasm.ts
import { createHash, randomUUID } from "node:crypto";
import { mkdtemp, mkdir, readFile, rm, writeFile } from "node:fs/promises";
import { closeSync, openSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { performance } from "node:perf_hooks";
import { deserialize, serialize } from "node:v8";
import { WASI } from "node:wasi";

import type {
  SandboxBackend,
  SandboxBackendCapabilities,
  SandboxExecutionRequest,
  SandboxExecutionResult,
} from "../types.js";

type WasmExecutorResult = {
  exitCode?: number;
  stdout?: string;
  stderr?: string;
};

export interface WasmCommandRegistration {
  modulePath: string;
  args?: string[];
  env?: Record<string, string>;
  preopens?: Record<string, string>;
}

export interface WasmSandboxBackendOptions {
  available?: boolean;
  executor?: (request: SandboxExecutionRequest) => Promise<WasmExecutorResult>;
  commands?: Record<string, WasmCommandRegistration>;
  logDir?: string;
  cacheDir?: string;
}

function normalizeArgs(args?: string[]): string[] {
  return Array.isArray(args) ? args.filter((value) => value.length > 0) : [];
}

function normalizeEnv(env?: Record<string, string>): Record<string, string> {
  if (!env) {
    return {};
  }

  return Object.fromEntries(
    Object.entries(env).filter((entry): entry is [string, string] => {
      const [key, value] = entry;
      return key.length > 0 && typeof value === "string";
    }),
  );
}

export class WasmSandboxBackend implements SandboxBackend {
  public readonly kind = "wasm" as const;
  private readonly commands: Map<string, WasmCommandRegistration>;
  private readonly moduleCache = new Map<string, WebAssembly.Module>();

  public constructor(private readonly options: WasmSandboxBackendOptions = {}) {
    this.commands = new Map(Object.entries(options.commands ?? {}));
  }

  public capabilities(): SandboxBackendCapabilities {
    return {
      kind: this.kind,
      supportsPersistence: false,
      supportsSnapshots: false,
      supportsBrowser: false,
      supportsNetwork: false,
      supportsPackageInstall: false,
      supportedCapabilities: ["command-exec"],
    };
  }

  public async isAvailable(): Promise<boolean> {
    return this.options.available ?? true;
  }

  public registerCommand(name: string, registration: WasmCommandRegistration): void {
    this.commands.set(name, registration);
  }

  public async execute(request: SandboxExecutionRequest): Promise<SandboxExecutionResult> {
    const startedAt = performance.now();

    try {
      const result = this.options.executor
        ? await this.options.executor(request)
        : await this.executeRegisteredModule(request);

      return {
        backend: this.kind,
        exitCode: result.exitCode ?? 0,
        stdout: result.stdout ?? "",
        stderr: result.stderr ?? "",
        durationMs: performance.now() - startedAt,
      };
    } catch (error) {
      return {
        backend: this.kind,
        exitCode: 1,
        stdout: "",
        stderr: error instanceof Error ? error.message : String(error),
        durationMs: performance.now() - startedAt,
      };
    }
  }

  private async loadModule(registration: WasmCommandRegistration): Promise<WebAssembly.Module> {
    const cached = this.moduleCache.get(registration.modulePath);
    if (cached) return cached;

    const moduleBytes = await readFile(resolve(registration.modulePath));

    if (this.options.cacheDir) {
      const hash = createHash("sha256").update(moduleBytes).digest("hex");
      const cachePath = join(this.options.cacheDir, `${hash}.v8cache`);
      try {
        const cacheData = await readFile(cachePath);
        const module = deserialize(cacheData) as WebAssembly.Module;
        this.moduleCache.set(registration.modulePath, module);
        return module;
      } catch {
        // cache miss — compile fresh
      }
      const module = await WebAssembly.compile(moduleBytes);
      await mkdir(this.options.cacheDir, { recursive: true });
      await writeFile(cachePath, serialize(module));
      this.moduleCache.set(registration.modulePath, module);
      return module;
    }

    const module = await WebAssembly.compile(moduleBytes);
    this.moduleCache.set(registration.modulePath, module);
    return module;
  }

  private async executeRegisteredModule(
    request: SandboxExecutionRequest,
  ): Promise<WasmExecutorResult> {
    const registration = this.commands.get(request.command);
    if (!registration) {
      throw new Error(`Unregistered WASI command: ${request.command}`);
    }

    const module = await this.loadModule(registration);

    let logDir: string;
    let persistent: boolean;

    if (this.options.logDir) {
      const ts = new Date().toISOString().replace(/[:.]/g, "-");
      const runId = randomUUID().slice(0, 8);
      logDir = join(this.options.logDir, `${ts}-${request.command}-${runId}`);
      await mkdir(logDir, { recursive: true });
      persistent = true;
    } else {
      logDir = await mkdtemp(join(tmpdir(), "titanclaw-wasi-"));
      persistent = false;
    }

    const stdoutPath = join(logDir, "stdout.log");
    const stderrPath = join(logDir, "stderr.log");
    const stdoutFd = openSync(stdoutPath, "w+");
    const stderrFd = openSync(stderrPath, "w+");

    try {
      const wasi = new WASI({
        version: "preview1",
        args: [
          request.command,
          ...normalizeArgs(registration.args),
          ...normalizeArgs(request.args),
        ],
        env: {
          ...normalizeEnv(registration.env),
          ...normalizeEnv(request.env),
        },
        preopens: registration.preopens ?? {},
        stdin: undefined,
        stdout: stdoutFd,
        stderr: stderrFd,
      });

      const imports = wasi.getImportObject() as WebAssembly.Imports;
      const instance = await WebAssembly.instantiate(module, imports);
      const start = (instance.exports as Record<string, unknown>)["_start"];
      if (typeof start !== "function") {
        throw new Error(`WASI module '${request.command}' does not export '_start'`);
      }

      let exitCode = 0;
      try {
        wasi.start(instance as WebAssembly.Instance);
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        const match = /exit code[: ]+(\d+)/i.exec(message);
        if (match) {
          exitCode = Number(match[1]);
        } else {
          throw error;
        }
      }

      closeSync(stdoutFd);
      closeSync(stderrFd);

      const stdout = await readFile(stdoutPath, "utf8");
      const stderr = await readFile(stderrPath, "utf8");
      return { exitCode, stdout, stderr };
    } finally {
      try {
        closeSync(stdoutFd);
      } catch {
        // ignore double-close
      }
      try {
        closeSync(stderrFd);
      } catch {
        // ignore double-close
      }
      if (!persistent) {
        await rm(logDir, { recursive: true, force: true });
      }
    }
  }
}

File: titanx/sandbox/backends/wasm.py
from __future__ import annotations

import asyncio
import hashlib
import os
import pickle
import tempfile
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Callable

from ..types import (
    SandboxBackend,
    SandboxBackendCapabilities,
    SandboxExecutionRequest,
    SandboxExecutionResult,
)


@dataclass
class WasmCommandRegistration:
    module_path: str
    args: list[str] = field(default_factory=list)
    env: dict[str, str] = field(default_factory=dict)
    preopens: dict[str, str] = field(default_factory=dict)


WasmExecutorFn = Callable[[SandboxExecutionRequest], Any]


class WasmSandboxBackend(SandboxBackend):
    kind = "wasm"

    def __init__(
        self,
        available: bool = True,
        executor: WasmExecutorFn | None = None,
        commands: dict[str, WasmCommandRegistration] | None = None,
        log_dir: str | None = None,
        cache_dir: str | None = None,
    ) -> None:
        self._available = available
        self._executor = executor
        self._commands: dict[str, WasmCommandRegistration] = commands or {}
        self._log_dir = log_dir
        self._cache_dir = cache_dir
        self._module_cache: dict[str, Any] = {}

    def capabilities(self) -> SandboxBackendCapabilities:
        return SandboxBackendCapabilities(
            kind="wasm",
            supports_persistence=False,
            supports_snapshots=False,
            supports_browser=False,
            supports_network=False,
            supports_package_install=False,
            supported_capabilities=["command-exec"],
        )

    async def is_available(self) -> bool:
        return self._available

    def register_command(self, name: str, reg: WasmCommandRegistration) -> None:
        self._commands[name] = reg

    async def execute(
        self,
        request: SandboxExecutionRequest,
        session=None,
    ) -> SandboxExecutionResult:
        start = time.perf_counter()
        try:
            if self._executor:
                result = await asyncio.get_event_loop().run_in_executor(
                    None, lambda: self._executor(request)  # type: ignore[misc]
                )
            else:
                result = await self._execute_registered(request)

            return SandboxExecutionResult(
                backend="wasm",
                exit_code=result.get("exit_code", 0),
                stdout=result.get("stdout", ""),
                stderr=result.get("stderr", ""),
                duration_ms=(time.perf_counter() - start) * 1000,
            )
        except Exception as exc:
            return SandboxExecutionResult(
                backend="wasm",
                exit_code=1,
                stdout="",
                stderr=str(exc),
                duration_ms=(time.perf_counter() - start) * 1000,
            )

    async def _execute_registered(self, request: SandboxExecutionRequest) -> dict:
        try:
            from wasmtime import Config, Engine, Linker, Module, Store, WasiConfig
        except ImportError:
            raise RuntimeError("wasmtime is not installed — pip install wasmtime")

        reg = self._commands.get(request.command)
        if not reg:
            raise ValueError(f"Unregistered WASI command: {request.command}")

        module = await self._load_module(reg.module_path)

        stdout_buf: list[str] = []
        stderr_buf: list[str] = []

        use_log_dir = self._log_dir is not None
        if use_log_dir:
            ts = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%S")
            run_dir = os.path.join(self._log_dir, f"{ts}-{request.command}")
            os.makedirs(run_dir, exist_ok=True)
            stdout_path = os.path.join(run_dir, "stdout.log")
            stderr_path = os.path.join(run_dir, "stderr.log")
        else:
            tmp = tempfile.mkdtemp(prefix="titanx-wasm-")
            stdout_path = os.path.join(tmp, "stdout.log")
            stderr_path = os.path.join(tmp, "stderr.log")

        def _run() -> dict:
            wasi = WasiConfig()
            wasi.argv = [request.command, *reg.args, *request.args]
            env_items = {**reg.env, **request.env}
            wasi.set_env(list(env_items.keys()), list(env_items.values()))
            for host_path, guest_path in reg.preopens.items():
                wasi.preopen_dir(host_path, guest_path)
            wasi.stdout_file = stdout_path
            wasi.stderr_file = stderr_path

            cfg = Config()
            engine = Engine(cfg)
            linker = Linker(engine)
            linker.define_wasi()
            store = Store(engine)
            store.set_wasi(wasi)

            instance = linker.instantiate(store, module)
            start_fn = instance.exports(store).get("_start")
            if start_fn is None:
                raise RuntimeError(f"WASI module '{request.command}' does not export '_start'")

            exit_code = 0
            try:
                start_fn(store)
            except Exception as exc:
                import re
                m = re.search(r"exit code[: ]+(\d+)", str(exc), re.I)
                if m:
                    exit_code = int(m.group(1))
                else:
                    raise

            with open(stdout_path) as f:
                stdout = f.read()
            with open(stderr_path) as f:
                stderr = f.read()

            return {"exit_code": exit_code, "stdout": stdout, "stderr": stderr}

        return await asyncio.get_event_loop().run_in_executor(None, _run)

    async def _load_module(self, module_path: str) -> Any:
        if module_path in self._module_cache:
            return self._module_cache[module_path]

        try:
            from wasmtime import Config, Engine, Module
        except ImportError:
            raise RuntimeError("wasmtime is not installed — pip install wasmtime")

        with open(module_path, "rb") as f:
            module_bytes = f.read()

        if self._cache_dir:
            digest = hashlib.sha256(module_bytes).hexdigest()
            cache_path = os.path.join(self._cache_dir, f"{digest}.pkl")
            if os.path.exists(cache_path):
                with open(cache_path, "rb") as f:
                    module = pickle.load(f)
                self._module_cache[module_path] = module
                return module
            engine = Engine(Config())
            module = Module(engine, module_bytes)
            os.makedirs(self._cache_dir, exist_ok=True)
            with open(cache_path, "wb") as f:
                pickle.dump(module, f)
        else:
            engine = Engine(Config())
            module = Module(engine, module_bytes)

        self._module_cache[module_path] = module
        return module
